### PR TITLE
Reintroduce build time init of compilation time metadata

### PR DESCRIPTION
**Pull request by graemerocher**
This PR reintroduces build time initialisation of compilation time metadata that was removed in 4.5.0

The changes allow Micronaut to support `--strict-image-heap` whilst still optimising for build time initialisation.

To achieve this evaluated expressions are now initialised at build time. This is safe because these cannot feature static initialisers. In addition enum introspections now no longer reference the enum values but lazily evaluate these.

### CODE REVIEW

File: core/src/main/java/io/micronaut/core/annotation/BuildTimeInit.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Annotate that stores references to class names that can be build time initialized inside bean definition metadata.
+ *
+ * <p>This annotation is currently designed for internal use only and not for public API.</p>
+ */
+@Experimental
+@UsedByGeneratedCode
+@Internal
+@Retention(RetentionPolicy.RUNTIME)
```
Comment (dstepanov) [inline core/src/main/java/io/micronaut/core/annotation/BuildTimeInit.java:31]: Make it `SOURCE` we don't need this annotation at runtime

**Review by dstepanov — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/micronaut/core/annotation/BuildTimeInit.java (hunk 1 above)
Comment (graemerocher) [reply core/src/main/java/io/micronaut/core/annotation/BuildTimeInit.java:31]: unfortunately `SOURCE` annotations cannot be read by native image features so don't know how to do this

Comment (dstepanov) [reply core/src/main/java/io/micronaut/core/annotation/BuildTimeInit.java:31]: Are you sure? Because that's using our own AnnotationMetadata

Comment (graemerocher) [reply core/src/main/java/io/micronaut/core/annotation/BuildTimeInit.java:31]: yes but the runtime metadata

Comment (graemerocher) [reply core/src/main/java/io/micronaut/core/annotation/BuildTimeInit.java:31]: if I change it to `SOURCE` the test fails

**Review by dstepanov — APPROVED** [review]

**Pull request merged**